File: backend/mystaticfiles/javascript/game/vec2.js
// import math

export class Vec2 {
	constructor ( x, y)
	{
		this.x = x
		this.y = y
	}


	print()
	{
		console.log("<{" + this.x + ":+.2f},{" + this.y + ":+.2f}>");
	}


	add( vec){
		this.x += vec.x
		this.y += vec.y
	}


	subBy( vec){
		this.x -= vec.x
		this.y -= vec.y
	}


	multiply( nb){
		this.x *= nb
		this.y *= nb
	}


	divide( nb){
		if (nb == 0)
			return
		this.x /= nb
		this.y /= nb
	}


	translate( x, y)
	{
		this.x += x
		this.y += y
	}


	translateAlong( vec, dist)
	{
		this.x += vec.x * dist
		this.y += vec.y * dist
	}


	rotateAround( x, y, sinTmp, cosTmp)
	{
		// Move point to center
		this.x -= x
		this.y -= y

		// Apply rotation
		tmpX = this.x
		tmpY = this.y
		this.x = (tmpX * cosTmp) - (tmpY * sinTmp)
		this.y = (tmpX * sinTmp) + (tmpY * cosTmp)

		// Uncenter point
		this.x += x
		this.y += y
	}


	asTupple()
	{
		return [this.x, this.y]
	}


	asTuppleCenter( x, y)
	{
		return [this.x - x, this.y - y]
	}


	norm()
	{
		return Math.sqrt(this.x**2 + this.y**2)
	}


	normalize()
	{
		let norm = this.norm()
		if (norm != 0){
			this.x /= norm
			this.y /= norm
		}
	}


	rotate( angle)
	{
		while (angle < 0)
			angle += 360
		while (angle > 359)
			angle -= 360

		let rad = angle * (Math.pi / 180)
		let cosRad = Math.cos(rad)
		let sinRad = Math.sin(rad)

		this.x = this.x * cosRad - this.y * sinRad
		this.y = this.x * sinRad + this.y * cosRad

		this.normalize()
	}


	dup()
	{
		return new Vec2(this.x, this.y)
	}

}

export function vec2Add(vec1, vec2)
{
	return new Vec2(vec1.x + vec2.x, vec1.y + vec2.y)
}


export function vec2Sub(vec1, vec2)
{
	return new Vec2(vec1.x - vec2.x, vec1.y - vec2.y)
}


export function vec2Dot(vec1, vec2)
{
	return (vec1.x * vec2.x) + (vec1.y * vec2.y)
}


export function vec2Cross(vec1, vec2)
{
	return (vec1.x * vec2.y) - (vec1.y * vec2.x)
}


export function getNormalOfSegment(vec1, vec2)
{
	dx = vec2.x - vec1.x
	dy = vec2.y - vec1.y
	vec = Vec2(-dy, dx)
	vec.normalize()
	return vec
}

export function reflectionAlongVec2(normal, vec)
{
	if (vec2Dot(normal, vec) >= 0)
		normal.multiply(-1)

	divider = vec2Dot(vec, vec)
	if (divider == 0)
		return (vec)
	vecProjOnNormal = normal.dup()
	vecProjOnNormal.multiply(vec2Dot(normal, vec) / divider)

	vecProjOnNormal.multiply(2)
	reflectedVec = vec2Sub(vec, vecProjOnNormal)

	reflectedVec.normalize()

	return reflectedVec
}

File: backend/pong_server/server_code/vec2.py
import math

class Vec2:
	def __init__(self, x, y) -> None:
		self.x = x
		self.y = y


	def __str__(self) -> str:
		return "<{x:+.2f},{y:+.2f}>".format(x=self.x, y=self.y)


	def add(self, vec) -> None:
		self.x += vec.x
		self.y += vec.y


	def subBy(self, vec) -> None:
		self.x -= vec.x
		self.y -= vec.y


	def multiply(self, nb) -> None:
		self.x *= nb
		self.y *= nb


	def divide(self, nb) -> None:
		if nb == 0:
			return
		self.x /= nb
		self.y /= nb


	def translate(self, x, y) -> None:
		self.x += x
		self.y += y


	def translateAlong(self, vec, dist) -> None:
		self.x += vec.x * dist
		self.y += vec.y * dist


	def rotateAround(self, x, y, sinTmp, cosTmp) -> None:
		# Move point to center
		self.x -= x
		self.y -= y

		# Apply rotation
		tmpX = self.x
		tmpY = self.y
		self.x = (tmpX * cosTmp) - (tmpY * sinTmp)
		self.y = (tmpX * sinTmp) + (tmpY * cosTmp)

		# Uncenter point
		self.x += x
		self.y += y


	def asTupple(self) -> tuple:
		return (self.x, self.y)


	def asTuppleCenter(self, x, y) -> tuple:
		return (self.x - x, self.y - y)


	def norm(self) -> float:
		return math.sqrt(self.x**2 + self.y**2)


	def normalize(self) -> None:
		norm = self.norm()
		if norm != 0:
			self.x /= norm
			self.y /= norm


	def rotate(self, angle) -> None:
		while angle < 0:
			angle += 360
		while angle > 359:
			angle -= 360

		rad = angle * (math.pi / 180)
		cosRad = math.cos(rad)
		sinRad = math.sin(rad)

		self.x = self.x * cosRad - self.y * sinRad
		self.y = self.x * sinRad + self.y * cosRad

		self.normalize()


	def dup(self):
		return Vec2(self.x, self.y)



def vec2Add(vec1, vec2) -> Vec2:
	return Vec2(vec1.x + vec2.x, vec1.y + vec2.y)


def vec2Sub(vec1, vec2) -> Vec2:
	return Vec2(vec1.x - vec2.x, vec1.y - vec2.y)


def vec2Dot(vec1, vec2) -> int:
	return (vec1.x * vec2.x) + (vec1.y * vec2.y)


def vec2Cross(vec1, vec2) -> int:
	return (vec1.x * vec2.y) - (vec1.y * vec2.x)


def getNormalOfSegment(vec1, vec2) -> Vec2:
	dx = vec2.x - vec1.x
	dy = vec2.y - vec1.y
	vec = Vec2(-dy, dx)
	vec.normalize()
	return vec


def reflectionAlongVec2(normal:Vec2, vec:Vec2) -> Vec2:
	if (vec2Dot(normal, vec) >= 0):
		normal.multiply(-1)

	divider = vec2Dot(vec, vec)
	if (divider == 0):
		return (vec)
	vecProjOnNormal = normal.dup()
	vecProjOnNormal.multiply(vec2Dot(normal, vec) / divider)

	vecProjOnNormal.multiply(2)
	reflectedVec = vec2Sub(vec, vecProjOnNormal)

	reflectedVec.normalize()

	return reflectedVec
